crates/spk-solve/src/io.rs

```diff
@@ -700,17 +733,128 @@ impl Pluralize for str {
+        match self {
+            OutputKind::Println => OutputKind::PrintlnAndToFile(output_file),
+            OutputKind::Tracing => OutputKind::TracingAndToFile(output_file),
+            OutputKind::LogFile(_) => OutputKind::LogFile(output_file),
+            OutputKind::PrintlnAndToFile(_) => OutputKind::PrintlnAndToFile(output_file),
+            OutputKind::TracingAndToFile(_) => OutputKind::TracingAndToFile(output_file),
+        }
+    }
+
+    /// Given a Println or Tracing output kind, this will ensure the
+    /// new output kind returned is one that includes Println or Tracing
+    /// respectively, as well as the output kind's existing file
+    /// output if any.
+    ///
+    /// This will error if any other output kind is given as the parameter.
+    fn include_output(self, other_output: &OutputKind) -> Result<Self> {
+        match other_output {
+            OutputKind::Println => match self {
+                OutputKind::Println => Ok(self.clone()),
+                OutputKind::Tracing => Err(Error::IncludingThisOutputNotSupported(
+                    "Cannot add Println output kind to a Tracing output kind. It".to_string(),
+                )),
+                OutputKind::LogFile(f) => Ok(OutputKind::PrintlnAndToFile(f.clone())),
+                OutputKind::PrintlnAndToFile(_) => Ok(self.clone()),
+                OutputKind::TracingAndToFile(_) => Err(Error::IncludingThisOutputNotSupported(
+                    "Cannot add Println output kind to a TrackingAndToFile output kind. It"
+                        .to_string(),
+                )),
+            },
+            OutputKind::Tracing => match self {
+                OutputKind::Println => Err(Error::IncludingThisOutputNotSupported(
+                    "Cannot add Tracing output kind to a Println output kind. It".to_string(),
+                )),
+                OutputKind::Tracing => Ok(self.clone()),
+                OutputKind::LogFile(f) => Ok(OutputKind::TracingAndToFile(f.clone())),
+                OutputKind::PrintlnAndToFile(_) => Err(Error::IncludingThisOutputNotSupported(
+                    "Cannot add Tracing output kind to a PrintlnAndToFile output kind. It"
+                        .to_string(),
+                )),
+                OutputKind::TracingAndToFile(_) => Ok(self.clone()),
+            },
+            _ => {
+                // OutputKinds other than Println or Tracing are not
+                // valid for combining with the current output kind here.
+                Err(Error::IncludingThisOutputNotSupported("OutputKind::ensure_output must be called with a Println or Tracing other_output. Including other kinds".to_string()))
+            }
+        }
+    }
+
+    /// Flush the output kind's output file, if any. Any errors that
+    /// occur while flushing the file are logged but not returned
+    /// because we don't want them to interrupt the solve.
+    fn flush(&self) {
```

Comment:
I'm not sure I agree with this approach because the reason an explicit flush exists is to surface any errors. But I do recognize this is a private method and logging a warning is likely the only way we want to proceed anyway.

If this was a public API I would want this to return `Result` so the caller can decide how to deal with errors.